p2p/src/cybershake.rs

```diff
@@ -286,6 +378,121 @@ impl<W: AsyncRead + Unpin> Incoming<W> {
     }
 }
 
+impl<W: AsyncRead + Unpin> Incoming<W> {
+    fn decipher_buf(&mut self) {
+        let seq = self.seq;
+        self.seq += 1;
+
+        self.kdf.append_u64(b"seq", seq);
+        let mut key = [0u8; 32];
+        self.kdf.challenge_bytes(b"key", &mut key);
+
+        let ad = encode_u64le(seq);
+
+        let siv_tag = GenericArray::clone_from_slice(&self.buf.as_slice()[..16]);
+        Aes128PmacSiv::new(GenericArray::clone_from_slice(&key))
+            .decrypt_in_place_detached(
+                &[&ad],
+                &mut self.buf.as_mut_slice()[16..self.ciphertext_length as usize],
+                &siv_tag,
+            )
+            .unwrap();
+
+        let pt_len = self.ciphertext_length as usize - 16;
+
+        for i in 0..pt_len {
+            let byte = self.buf.as_slice()[i + 16];
+            self.buf.as_mut_slice()[i] = byte;
+        }
+
+        self.plaintext_length = pt_len;
+        self.ciphertext_length = 0;
+        self.ciphertext_read = 0;
+    }
+
+    fn read_buffer(&mut self, buf: &mut [u8]) -> Result<usize, io::Error> {
+        match Read::read(
+            &mut &self.buf[self.plaintext_read..self.plaintext_length],
+            buf,
+        ) {
+            Ok(n) => {
+                self.plaintext_read += n;
+                if self.plaintext_read == self.plaintext_length {
+                    self.plaintext_read = 0;
+                    self.plaintext_length = 0;
+                }
+                Ok(n)
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl<W: AsyncRead + Unpin> AsyncRead for Incoming<W> {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let me = self.get_mut();
+
+        if me.plaintext_length != 0 {
+            return Poll::Ready(me.read_buffer(buf));
+        }
+
+        if me.ciphertext_length == 0 {
+            loop {
+                let poll = me
+                    .reader
+                    .as_mut()
+                    .poll_read(cx, &mut me.buf[me.ciphertext_read as usize..2]);
```

Comment:
This line calls for a better code organization: currently it reads "we may have read some bytes, but we want 2 more", which safe in the overall context (because we expect 16 bytes of auth tag anyway), but locally does not look correct. Maybe better to have an enum for reading that encodes states properly:

enum ReadState{
Len(n:usize) // contains number of bytes read so far. We need to read `2-n`.
Body(n: usize, k:usize) // total length, and number of _body_ bytes read, ignoring the length prefix that we already have.
}

Reply:
Then we don't need `ciphertext_read`/`ciphertext_length` variables. 

When we are done, we revert to Len(0) state (aka "expecting to read the length prefix, but have 0 bytes so far").

Reply:
Added enum ReadState